distributed/tests/test_scheduler.py

```diff
@@ -2442,3 +2443,23 @@ async def test_memory_no_zict(c, s, a, b):
 async def test_memory_no_workers(s):
     assert s.memory.process == 0
     assert s.memory.managed == 0
+
+
+@gen_cluster(client=True, nthreads=[])
+async def test_memory_is_none(c, s):
+    """If Worker.heartbeat() runs before Worker.monitor.update(), then
+    Worker.metrics["memory"] will be None and will need special handling in
+    Worker.memory and Scheduler.heartbeat_worker().
+    """
+    with mock.patch("distributed.system_monitor.SystemMonitor.update"):
```

Comment:
Do either of @fjetter's suggestions in https://github.com/dask/distributed/pull/4719#discussion_r617537532 allow us to test the case where the worker heartbeats before the system monitor has updated? It'd be nice to avoid `mock`ing if possible

Reply:
Alright, thanks for trying them (and a couple other options) out 👍